File: app/src/main/java/com/example/dataLayer/repositories/PostRepository.java
package com.example.dataLayer.repositories;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.bookapp.models.Post;
import com.example.dataLayer.dataObjectsToSerialize.SerializePost;
import com.example.dataLayer.interfaces.PostRepositoryInterface;

import java.util.ArrayList;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class PostRepository {

    //these fields should only represent caching data

    private static PostRepository instance;
    private MutableLiveData<ArrayList<Post>> posts;
    private MutableLiveData<Post> currentFetchedPost;
    private MutableLiveData<ArrayList<Post>> favoritePosts;
    private MutableLiveData<ArrayList<Post>> myPosts;
    private PostRepositoryInterface repositoryInterface;
    private MutableLiveData<Post> currentlyUploadedPost = new MutableLiveData<>();
    private MutableLiveData<ArrayList<Post>> newFetchedPosts  = new MutableLiveData<>();


    private PostRepository(@NonNull Retrofit retrofit) {
        repositoryInterface = retrofit.create(PostRepositoryInterface.class);
    }


    public MutableLiveData<Post> getCurrentlyUploadedPost() {
        return currentlyUploadedPost;
    }

    public static PostRepository getInstance(@NonNull Retrofit retrofit) {
        if (instance == null) {
            instance = new PostRepository(retrofit);
        }
        return instance;

    }

    public MutableLiveData<ArrayList<Post>> fetchFirstPagePosts() {
        posts = new MutableLiveData<>();
        repositoryInterface.fetchPosts(true).enqueue(new Callback<ArrayList<Post>>() {
            @Override
            public void onResponse(@NonNull Call<ArrayList<Post>> call, @NonNull Response<ArrayList<Post>> response) {
                ArrayList<Post> fetchedPosts = response.body();
                assert fetchedPosts != null;
                if (posts.getValue() != null) {
                    //clear some data if necessary
                    clearData(posts);
                    posts.getValue().addAll(fetchedPosts);
                } else {
                    posts.setValue(fetchedPosts);
                }
            }

            @Override
            public void onFailure(@NonNull Call<ArrayList<Post>> call, Throwable t) {

            }
        });
        return posts;
    }

    private void clearData(@NonNull MutableLiveData<ArrayList<Post>> posts) {
    }

    public MutableLiveData<Post> fetchPostByID(long id) {
        currentFetchedPost = new MutableLiveData<>();
        repositoryInterface.fetchPostByID(id).enqueue(new Callback<Post>() {
            @Override
            public void onResponse(Call<Post> call, Response<Post> response) {
                currentFetchedPost.setValue(response.body());
                Log.d("test", "callback with uploaded data");
            }

            @Override
            public void onFailure(Call<Post> call, Throwable t) {

            }
        });
        return currentFetchedPost;
    }

    /**
     * This method should be called when the favorite posts
     * data is requested
     * The method decided weather it should fetch the data from
     * cache or from the source
     *
     * @param userID
     * @return
     */
    public MutableLiveData<ArrayList<Post>> fetchFavoritePosts(String userID) {
        //prepare data to be returned on the main thread
        favoritePosts = new MutableLiveData<>();
        //start fetching the other data on the other thread
        repositoryInterface.fetchFavoritePostsByUserID(userID, true).enqueue(new Callback<ArrayList<Post>>() {
            @Override
            public void onResponse(Call<ArrayList<Post>> call, Response<ArrayList<Post>> response) {
                ArrayList<Post> fetchedPosts = response.body();
                fetchedPosts.forEach(post -> post.setFavorite(true));
                favoritePosts.setValue(fetchedPosts);
            }

            @Override
            public void onFailure(Call<ArrayList<Post>> call, Throwable t) {

            }
        });
        return favoritePosts;
    }

    public MutableLiveData<ArrayList<Post>> fetchMyPosts(String userID) {
        myPosts = new MutableLiveData<>();
        repositoryInterface.fetchMyPosts(userID, true).enqueue(new Callback<ArrayList<Post>>() {
            @Override
            public void onResponse(Call<ArrayList<Post>> call, Response<ArrayList<Post>> response) {
                myPosts.setValue(response.body());
            }

            @Override
            public void onFailure(Call<ArrayList<Post>> call, Throwable t) {

            }
        });
        return myPosts;
    }


    public void addPostToFavorites(long postID, String userID) {
        repositoryInterface.addPostToFavorites(true, postID, userID).enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull Call<String> call, @NonNull Response<String> response) {

            }

            @Override
            public void onFailure(@NonNull Call<String> call, @NonNull Throwable t) {

            }
        });
    }

    public void insertPost(@NonNull SerializePost serializePost) {

        repositoryInterface.uploadPost(true, serializePost).enqueue(new Callback<Post>() {
            @Override
            public void onResponse(@NonNull Call<Post> call, @NonNull Response<Post> response) {
                currentlyUploadedPost.setValue(response.body());
            }

            @Override
            public void onFailure(@NonNull Call<Post> call, @NonNull Throwable t) {
                t.printStackTrace();
            }
        });
    }




    public void deletePostFromFavorites(long postID, String userID) {
        //  repositoryInterface.deletePostFromFavorites(postID,userID);
    }


    public void fetchNewPosts() {
        Log.d("HomeFragment","fetching new posts");
        repositoryInterface.fetchPosts(true).enqueue(new Callback<ArrayList<Post>>() {
            @Override
            public void onResponse(@NonNull Call<ArrayList<Post>> call, @NonNull Response<ArrayList<Post>> response) {
                newFetchedPosts.setValue(response.body());
            }

            @Override
            public void onFailure(@NonNull Call<ArrayList<Post>> call, Throwable t) {

            }
        });

    }

    public MutableLiveData<ArrayList<Post>> getNewFetchedPosts() {
        return newFetchedPosts;
    }
}

File: app/src/main/java/com/example/dataLayer/repositories/PostRepository.kt
package com.example.dataLayer.repositories

import android.net.Uri
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.liveData
import androidx.paging.PagedList
import com.example.bookapp.models.Post
import com.example.bookapp.models.User
import com.example.dataLayer.dataMappers.toPost
import com.example.dataLayer.interfaces.PostRepositoryInterface
import com.example.dataLayer.interfaces.dao.RoomPostDao
import com.example.dataLayer.models.*
import com.example.dataLayer.models.serialization.SerializePost
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.storage.FirebaseStorage
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.InternalCoroutinesApi
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import javax.inject.Inject


@InternalCoroutinesApi

class PostRepository @Inject constructor(private val user: User,
                                         private val requestExecutor: RequestExecutor,
                                         private val coroutineScope: CoroutineScope,
                                         private val repo: PostRepositoryInterface,
                                         private val firebaseRepo: FirebaseFirestore,
                                         private val firebaseStorage: FirebaseStorage,
                                         private val postDao: RoomPostDao
) {

    private val postCollection = "posts"

    fun getPosts() = postDao.getCachedPosts().also {
        coroutineScope.launch {
            //if network is active remove old data and
            //perform a fresh fetch
            fetchInitialPosts()

            requestExecutor.add(this@PostRepository::fetchFavoritePostsImpl, null)

        }
    }


    inner class PostRepoBoundaryCallback : PagedList.BoundaryCallback<Post>() {
        override fun onZeroItemsLoaded() {
            //when no items were loaded from room ,trigger a network call
            coroutineScope.launch {
                fetchInitialPosts()
            }
        }

        override fun onItemAtEndLoaded(itemAtEnd: Post) {
            coroutineScope.launch {
                //  fetchNextPosts(itemAtEnd.id)
            }
        }
    }

    val favoritePosts: LiveData<UserWithFavoritePosts> by lazy {
        liveData(Dispatchers.IO) {
            emitSource(postDao.getFavoritePosts(user.userID))
            requestExecutor.add(this@PostRepository::fetchFavoritePostsImpl, null)
        }
    }


    fun fetchPostByID(id: String): LiveData<Post> = liveData {
        val fetchedData = firebaseRepo.collection(postCollection).document(id).get().await().toObject(PostDTO::class.java)
        fetchedData?.let {
            val post = it.toPost()
            postDao.insertPost(post)
            emit(post)
        }
    }


    internal suspend fun fetchFavoritePostsImpl() {
        val data =
                repo.fetchUserFavoritePosts(user.userID).map { it.toPost() }

//        postDao.insertAllFavoritePosts(data.map {
//            UserWithFavoritePostsCrossRef(postID = it.id, userID = user.userID)
//        })
    }


    fun fetchMyPosts() = liveData {
        emitSource(postDao.getAllUserPosts(user.userID))
        try {
            val fetchedPosts = repo.fetchMyPosts(user.userID)
            postDao.insertPosts(fetchedPosts.map { it.toPost() })
        } catch (e: java.lang.Exception) {
            e.printStackTrace();
        }
    }

    suspend fun addPostToFavorites(post: Post) {
        //     postDao.addFavoritePost(UserWithFavoritePostsCrossRef(postID = post.id, userID = user.userID))
        //     repo.addPostToFavorites(post.id, user.userID)
    }


    suspend fun deletePostFromFavorites(post: Post) {
        // repo.removePostFromFavorites(postID = post.id, userID = user.userID)
        //  val toRemove = UserWithFavoritePostsCrossRef(postID = post.id, userID = user.userID)
        //  postDao.deletePostFromFavorites(toRemove)

    }


    suspend fun fetchInitialPosts() {
        postDao.removeCachedData()
        firebaseRepo.collection(postCollection).orderBy(Post::date.name, Query.Direction.DESCENDING).limit(10)
                .addSnapshotListener { snapshots, e ->
                    if (e != null) {
                        return@addSnapshotListener
                    }
                    snapshots?.let {
                        val data = snapshots.mapNotNull {
                            Log.d("test", it.id)
                            it.toObject(PostDTO::class.java).toPost(it.id)
                        }
                        coroutineScope.launch { postDao.insertPosts(data) }
                    }
                }
    }

    internal suspend fun fetchNextPosts(lastPostID: Int) {
        try {
            val fetchedData = repo.fetchNextPagePosts(lastPostID)
            postDao.insertPosts(fetchedData.map { it.toPost() })
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
    }

    fun uploadImage(serializeImage: SerializeImage): LiveData<String> =
            liveData {
                emit(String())
                val imagePath = repo.uploadImage(serializeImage).message
                emit(imagePath)
            }

    fun uploadPost(postDTO: PostDTO): LiveData<OperationStatus> {
        return liveData {
            //mark the operation ongoing
            emit(OperationStatus.ONGOING)
            //upload the post
            val id = firebaseRepo.collection(postCollection).add(postDTO).await().id
            val fetchedData = firebaseRepo.collection(postCollection).document(id).get().await().toObject(PostDTO::class.java)
            fetchedData?.let {
                val post = it.toPost(id)
                postDao.insertPost(post)
                emit(OperationStatus.FINISHED)
            }

        }
    }

    fun uploadFirebaseImage(path: Uri): LiveData<String> = liveData {
        emit("")
        val filePath = "images/posts/${path.lastPathSegment}"
        firebaseStorage.reference.child(filePath).putFile(path).await()
        //fetch the full url
        val fullUrl = firebaseStorage.getReference(filePath).downloadUrl.await().toString()
        emit(fullUrl)
    }
}



